design-proposals/2023-02-08-code-organisation.md

```diff
@@ -0,0 +1,137 @@
+
+The code should be C++17 or later (unless required otherwise by platform-specific integration).
+
+Compilation flags should ideally be the same on one native platform / operating system. This would allow building multiple platform-specific parts in parallel and would speed-up the development.
+
+Final products of the core would be two or more object libraries to be directly used by the SDKs or one static or dynamic library for binary distribution. *Note: At this point it is not clear if all platforms need binary distribution independent of the SDK that will be provided for that platform.*
+
+Examples of binary distributions could be:
+
+- Dynamic libraries for Windows, Linux and Android.
+- Dynamic and static frameworks for macOS and iOS.
+- Cross-platform engines will probably build the core as part of the SDK and no direct core binaries are needed (as written above this may also apply for native platforms).
+
+### Testing
+
+All testing should be done for all platforms at the same time and as complete as possible. Basic rendering tests should be done headlessly for all platforms.
+
+If needed for specific platforms (iOS comes to mind), a dedicated test app could be made to run the necessary rendering tests. This app should implement the minimal functionality that does not include the actual full-fledged integration of MapLibre (e.g. gestures, widgets, ...).
+
+Testing should be done by `CTest` if the platform allows (may not be possible for all integration tests on mobile devices).
+
+### Release cycle
+
+The core should have an independent release cycle. At this point we imagine SDKs to be much more actively developed. This also allows bigger refactoring to take place without affecting SDKs (e.g. Metal migration).
+
+## SDKs and other bindings
+
+SDKs, language bindings and platform bindings should provide utilities to integrate MapLibre better to a specific platform. They should not add any core mapping functionality but they could add integration utilities that can not be made platform independent, e.g.
+
+- widgets
+- gestures support on toucscreen devices
+- location services (GPS, location from the mobile network)
+
+### Organisation
+
+MapLibre core could be provided as a submodule, but it is not necessary. For mobile platforms this may or may not be the case, depending on the level of the integration, for example:
+
+- on Android if Kotlin/Java is mostly used, then distributing the core separately as a dynamic library is completely reasonable
+- on iOS where static building may be much more beneficial due to size the core will probably be built together with the SDK
```

Comment:
@1ec5 mentioned on Slack that a source-only distribution is the prefered way of distribution for Apple platforms now. For MapLibre I don't know if would apply to just the SDK (probably written in Swift in the long term) or also the core written in C++.

Reply:
This was in the context of the discussion on package managers. It’s a lot easier to set up, for instance, SPM based on source distribution. MapLibre already has a workaround for SPM based on a binary distribution in [a dedicated repository](https://github.com/maplibre/maplibre-gl-native-distribution), but it isn’t ideal in terms of requiring additional release steps and less obvious installation steps.

There are some hurdles to source-only distribution: the CMake-based build system is not straightforward enough for tight integration with package managers. mbgl takes a relatively long time to build. The whole gl-native repository and its submodules are too large to comfortably make every application developer clone it anew every time they resolve dependencies.

These hurdles mostly go away if the platform-specific code (other than a few low-level odds and ends) wind up in a separate repository like mapbox-gl-native-ios. That repository couldn’t achieve source-only distribution partly because it carried the entire commit and tag history of gl-native along with it: https://github.com/maplibre/maplibre-gl-native/pull/789#discussion_r1100558480.

To be clear, source distribution is not mutually exclusive of binary distribution. It would remain possible to attach a prebuilt binary to a GitHub release, just as the gl-native-distribution repository currently does.

Reply:
Onboard with the package manager point.

Reply:
By the way, if core is packaged up as a build artifact for iOS or macOS, note that a framework can’t expose any Objective-C++ because the C++ won’t translate across ABI boundaries. This means any framework/library vended by core for iOS/macOS would need to stick to pure C or C++ in its public interface. This should already be the case, but it would influence where we draw the line between core and SDK codebases.